app/frontend/actions/send.ts

```diff
@@ -0,0 +1,307 @@
+        address: newState.sendAddress.fieldValue as Address,
+        coins,
+        token,
+        minimalLovelaceAmount: txPlanResult.txPlan.additionalLovelaceAmount,
+      }
+      setTransactionSummary(txPlanResult.txPlan, sendTransactionSummary)
+      setState({
+        calculatingFee: false,
+        txSuccessTab: '',
+        transactionFee: txPlanResult.txPlan.fee,
+      })
+    } else {
+      /*
+      REFACTOR: (calculateFee)
+      Handling validation error should not be the responsibility of action called "calculateFee"
+      */
+      const validationError =
+        txPlanValidator(
+          coins,
+          txPlanResult.minimalLovelaceAmount,
+          balance,
+          txPlanResult.estimatedFee
+        ) || txPlanResult.error
+      setError(state, {errorName: 'sendAmountValidationError', error: validationError})
+      setState({
+        calculatingFee: false,
+        txSuccessTab: '',
+        transactionFee: txPlanResult.estimatedFee,
+      })
+    }
+  }
+
+  const debouncedCalculateFee = debounceEvent(calculateFee, 2000)
+
+  /*
+  REFACTOR: (forms)
+  This logic & state should be moved to components.
+  */
+  const validateSendFormAndCalculateFee = () => {
+    validateSendForm(getState())
+    resetTransactionSummary(getState())
+    setState({transactionFee: 0})
+    const state = getState()
+    if (isSendFormFilledAndValid(state)) {
+      setState({calculatingFee: true})
+      debouncedCalculateFee()
+    } else {
+      setState({calculatingFee: false})
+    }
+  }
+
+  /*
+  REFACTOR: (forms)
+  This logic & state should be moved to components.
+  */
+  const updateAddress = (state: State, e, address?: string) => {
+    setState({
+      sendResponse: '',
+      sendAddress: Object.assign({}, state.sendAddress, {
+        fieldValue: address || e.target.value,
+      }),
+    })
+    validateSendFormAndCalculateFee()
+  }
+
+  /*
+  REFACTOR: (forms)
+  This logic & state should be moved to components.
+  */
+  const updateAmount = (state: State, sendAmount: SendAmount): void => {
+    setState({
+      sendResponse: '',
+      sendAmount: Object.assign({}, state.sendAmount, sendAmount),
+    })
+    validateSendFormAndCalculateFee()
+  }
+
+  const validateAndSetMaxFunds = (state: State, maxAmount: SendAmount) => {
+    // TODO: some special validation
+
+    updateAmount(state, maxAmount)
+  }
+
+  const sendMaxFunds = async (state: State) => {
+    setState({calculatingFee: true})
+    try {
+      const maxAmounts = await getWallet()
+        .getAccount(state.sourceAccountIndex)
+        .getMaxSendableAmount(state.sendAddress.fieldValue as Address, state.sendAmount)
+      validateAndSetMaxFunds(state, maxAmounts)
+    } catch (e) {
+      setState({
+        calculatingFee: false,
+      })
+      setError(state, {errorName: 'sendAmountValidationError', error: {code: e.name}})
+      return
+    }
+  }
+
+  const convertNonStakingUtxos = async (state: State): Promise<void> => {
```

Comment:
I am not sure whether this is the best place for this and `witdrawRewards` action, but I do not better alternatives right now. Someone else any ideas?

Reply:
I moved it to `actions/transaction.ts` for now.